packages/core/src/app/pg-core/db/chain-db.ts

```diff
@@ -1,24 +1,523 @@
+    if (node === null) {
+      return
+    }
+
+    // We have the node value, but we don't know whether we actually have the
+    // value stored or if we have children that can re-generate the node.
+
+    // First, check if we have the node in question.
+    const key = KEYS.TREE_NODES.encode([blockNumber, nodeIndex])
+    const value = await this.db.get(key)
+
+    // We have this specific node, so operate on it.
+    if (value !== null) {
+      // Delete the node.
+      await this.db.del(key)
+
+      // Figure out if this node had a sibling. If so, we're going to want to
+      // compute the parent and insert it since we'd be losing that info.
+      const siblingIndex = getSiblingIndex(nodeIndex)
+      const sibling = await this.getMerkleTreeNode(blockNumber, siblingIndex)
+
+      // We have a sibling, so we're going to compute the parent and insert it.
+      if (sibling !== null) {
+        const parentIndex = getParentIndex(nodeIndex)
+        const parent = isLeftChild(nodeIndex)
+          ? computeParent(node, sibling)
+          : computeParent(sibling, node)
+        await this.addMerkleTreeNode(blockNumber, parentIndex, parent)
+      }
+
+      // Stop, don't need to delete anything else.
+      return
+    }
+
+    // We don't have the node, but we have children that can be used to
+    // re-generate it. Need to recursively delete any of these children.
+
+    // We're at a leaf node so we can't go any deeper.
+    const treeHeight = await this.getBlockTreeHeight(blockNumber)
+    if (nodeIndex > 2 ** treeHeight) {
+      return
+    }
+
+    // Compute left and right children.
+    const leftChildIndex = getLeftChildIndex(nodeIndex)
+    const rightChildIndex = getRightChildIndex(nodeIndex)
+
+    // Delete the children.
+    await this.removeMerkleTreeNode(blockNumber, leftChildIndex)
+    await this.removeMerkleTreeNode(blockNumber, rightChildIndex)
+  }
+
+  /**
+   * Smart method for removing Merkle proof nodes stored for a specific
+   * transaction. Checks the ensure that deletion of the nodes wouldn't impact
+   * other transaction proofs.
+   * @param blockNumber Block number to delete nodes from.
+   * @param leafIndex Index of the leaf to delete proof nodes for.
+   */
+  public async removeMerkleProofNodes(
+    blockNumber: number,
+    leafIndex: number
+  ): Promise<void> {
+    // Get the leaf indices of all other transactions in the same block.
+    const otherLeafIndices: number[] = (await this.getAllLeafIndices(
+      blockNumber
+    )).filter((otherLeafIndex) => {
+      return otherLeafIndex !== leafIndex
+    })
+
+    // Need the tree height to compute sibling/parent nodes.
+    const treeHeight = await this.getBlockTreeHeight(blockNumber)
+
+    // We don't want to delete any nodes that other transactions still need.
+    // Compute the siblings for *all* other stored transactions so we're not
+    // deleting anything critical.
+    let allOtherSiblingIndices: number[] = []
```

Comment:
Makes sure we don't delete any nodes that *other* transactions in the same block still need.